spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -258,24 +270,32 @@ trait TableFeatureSupport { this: Protocol =>
   def canDowngradeTo(to: Protocol, droppedFeatureName: String): Boolean = {
     if (!supportsWriterFeatures) return false
 
-    // When `to` protocol does not have any features version downgrades are possible. However,
+    val featureNames = readerAndWriterFeatureNames - droppedFeatureName
+    val minRequiredVersions = TableFeatureProtocolUtils.minimumRequiredVersions(
+      featureNames.flatMap(TableFeature.featureNameToFeature).toSeq)
+    val toSupportsRequiredMinVersions =
+      (to.minReaderVersion, to.minWriterVersion) == minRequiredVersions
+
+    // When `to` protocol does not have any features, versions might get downgraded. However,
     // the current protocol needs to contain one non-legacy feature. We also allow downgrade when
     // there are only legacy features. This is to accommodate the case when the user attempts to
     // remove a legacy feature in a table that only contains legacy features.
     if (to.readerAndWriterFeatureNames.isEmpty) {
-      val featureNames = readerAndWriterFeatureNames - droppedFeatureName
       val sameLegacyFeaturesSupported = featureNames == to.implicitlySupportedFeatures.map(_.name)
-      val minRequiredVersions = TableFeatureProtocolUtils.minimumRequiredVersions(
-        featureNames.flatMap(TableFeature.featureNameToFeature).toSeq)
 
       return sameLegacyFeaturesSupported &&
         (to.minReaderVersion, to.minWriterVersion) == minRequiredVersions &&
         readerAndWriterFeatures.filterNot(_.isLegacyFeature).size <= 1
     }
 
-    // When `to` protocol contains table features we cannot downgrade the protocol version.
-    if (to.minReaderVersion != this.minReaderVersion) return false
-    if (to.minWriterVersion != this.minWriterVersion) return false
+    // When `to` protocol contains table features, protocol versions may still
```

Comment:
I find this very hard to understand because we're handling so many cases separately. I think it's organically grown into this, but basically all we're trying to say is "the new protocol should be the old protocol minus the one feature that we're dropping". I think the entirety of this function can be simplified into something as stupid as:

```
(this.implicitlyAndExplicitlySupportedFeatures -- to.implicitlyAndExplicitlySupportedFeatures).toSeq == Seq(TableFeature.featureNameToFeature(droppedFeature))
```

I know that there's the point below that "Note, we check the minimum required versions only when table features exist in the protocol." But I think that's a choice that gets made at a different level, where we calculate the "to" protocol that we're testing against. But the basic test for "is this a valid downgrade" can be very simple.

Reply:
Yes that is great. I will just turn it a bit around because like this it requires a symmetric difference.